### PR TITLE
Focus Control.

### PR DESCRIPTION
**Pull request by andersio**
This PR provides an infrastructure of focus coordination among Bento components.

### Creating a focusable component
For a Bento component to participate in the focus coordination, it would need to:

1. Conform to `Focusable`.

    Components may specify whether they participate in the focus coordination via `canBeFocused`. This is important for the focus coordination to discover all focusable components, especially those that are off screen.

2. Have its view conforming to `FocusableView`.

    The focus coordination invokes `focus()` on the conforming component root view when appropriate. The root view should respond [by asking the appropriate subviews (or itself) to assume the first responder status](https://github.com/Babylonpartners/Bento/blob/f00daad852eb02af7ecaa4a9d73815929da1e0f0/Example/Cells/TextFieldComponent.swift#L63).

    Conforming root views would automatically gain `withFocusCoordinator(_:)`, which provides scoped access to the focus coordination mechanism. The coordinator can be used to:

    a. [attempt to pass on its first responder status via `move(_:)`](https://github.com/Babylonpartners/Bento/blob/f00daad852eb02af7ecaa4a9d73815929da1e0f0/Example/Cells/TextFieldComponent.swift#L80); or

    b. [query whether there is any focusable component following it via `canMove(_:)`, so as to adjust the keyboard return key type, for example](https://github.com/Babylonpartners/Bento/blob/f00daad852eb02af7ecaa4a9d73815929da1e0f0/Example/Cells/TextFieldComponent.swift#L69).

3. Optionally be notified when a new 🍱 has been rendered.

    For example, if you have a input accessory toolbar with buttons to navigate between form fields, the availability of the buttons might be invalidated by new 🍱 boxes & therefore need to be updated. One may implement `neighbouringFocusEligibilityDidChange()` to receive such messages.

### Focus without user interaction

Bento introduces `focus(animated:)` to `UITableView` and `UICollectionView`, which upon invocation would scroll to the first focusable component & ask it to focus.

### Example: A personal details form.
Available as the fifth tab in the Bento example app. Check `PersonalDetailsViewController.swift` for the source.

### Implementation Note

This is architecturally the same as the focus control in our pre-Bento infrastructure. The notable difference is the implementation being extended to support both kind of collection views.

`FocusCoordinating` is a transient view of the abstract focus coordination mechanism.

It is exposed as a scoped accessor `withFocusCoordinator(_:)` to discourage escaping/retaining the coordinator. Prior Art: `withoutActuallyEscaping(_:)`.

Bento component views have zero knowledge about its position within the collection view. So for the purpose of focus coordination we need to query it & have it stored temporarily, but the info should be discarded as soon as we are done with the coordination.

The implementations introduce a group of internal protocols, centred by `BentoCollectionView`. `BentoContainerView` defines a set of requirements that is necessary for focus coordination (implemented as an extension to `BentoCollectionView`). Auxiliary protocols are introduced to either provide stronger constraints for container-agnostic implementations, or have a subset that can be used as existentials (protocol types) in dynamic type casting.

### CODE REVIEW

File: Bento/Helpers.swift [hunk 1]
```diff
@@ -63,3 +63,11 @@ extension Optional {
         return selector(unwrapped, other)
     }
 }
+
+internal func search<T>(from leaf: UIView, type: T.Type) -> T? {
+    var cell = leaf.superview
+    while let view = cell, !(view is T) {
+        cell = view.superview
+    }
+    return cell as! T?
```
Comment (sergdort) [inline Bento/Helpers.swift]: Why not `cell as? T`?

Comment (andersio) [reply Bento/Helpers.swift]: This is an explicit assertion. It must be `T` or `nil` at this point.

File: Bento/Adapters/BentoCollectionView.swift [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    var dataSource: DataSource? { get }
+
+    func indexPath(for cell: Cell) -> IndexPath?
+    func visibleCell(at indexPath: IndexPath) -> Cell?
+    func revealCell(at indexPath: IndexPath, animated: Bool)
+}
+
+extension BentoCollectionView {
+    func didRenderBox() {
+        UIApplication.shared.sendAction(#selector(FocusableView.neighbouringFocusEligibilityDidChange),
```
Comment (mluisbrown) [inline Bento/Adapters/BentoCollectionView.swift]: There's something so cool about being able to fire a message all the way up the responder chain 🚀

File: Bento/Renderable/AnyRenderable.swift [hunk 3]
```diff
@@ -1,14 +1,10 @@
...
-struct AnyRenderable: Renderable, Deletable {
-    var canBeDeleted: Bool {
-        return base.canBeDeleted
-    }
+protocol ExistentialWrapper {
 
-    var deleteActionText: String {
-        return base.deleteActionText
-    }
+}
```
Comment (mluisbrown) [inline Bento/Renderable/AnyRenderable.swift]: [nit] extra line above. Actually since this is a marker protocol, could just be `{ }`

File: Bento/Renderable/FocusableView.swift [hunk 4]
```diff
@@ -0,0 +1,25 @@
+@objc(BentoFocusableView)
+public protocol FocusableView: AnyObject {
+    @objc(bento_focus)
+    func focus()
+
+    @objc(bento_neighbouringFocusEligibilityDidChange)
+    optional func neighbouringFocusEligibilityDidChange()
```
Comment (mluisbrown) [inline Bento/Renderable/FocusableView.swift]: This needs to be documented. I don't fully understand it's use :/

[nit] possible should use US spelling (`neighboring`) to match Apple frameworks... 😷

Comment (andersio) [reply Bento/Renderable/FocusableView.swift]: Done.

File: Bento/Renderable/Deletable.swift [hunk 5]
```diff
@@ -15,3 +13,15 @@ public extension Deletable {
         return nil
     }
 }
+
+// NOTE: Workaround to the lack of generalized existentials for Protocols with
+//       Associated Types (PATs).
+//
+//       See `_Focusable` for the rationale.
+/// Bento internal protocol. See `Deletable` for the public API.
+public protocol _Deletable {
```
Comment (sergdort) [inline Bento/Renderable/Deletable.swift]: Why is this changed? 
It was working fine before no?

File: Bento/Adapters/SectionedFormAdapter.swift [hunk 6]
```diff
@@ -97,11 +99,12 @@ final class SectionedFormAdapter<SectionId: Hashable, RowId: Hashable>
     @available(iOS 11.0, *)
     func tableView(_ tableView: UITableView, trailingSwipeActionsConfigurationForRowAt indexPath: IndexPath) -> UISwipeActionsConfiguration? {
         let row = sections[indexPath.section].rows[indexPath.row]
-        guard row.component.canBeDeleted else {
+        guard let component = row.component.cast(to: _Deletable.self),
```
Comment (sergdort) [inline Bento/Adapters/SectionedFormAdapter.swift]: What was wrong with default implementation?

File: Bento/Renderable/Deletable.swift (hunk 5 above)
Comment (andersio) [reply Bento/Renderable/Deletable.swift]: I want to keep `Focusable` inheriting from `Renderable` as `Deletable` does. However, with the existing configuration it would require two extra existential wrappers (`AnyFocusableBox` and `AnyFocusableDeletableBox`), and even more stuff being dumped into `AnyRenderable`. This isn't a scalable solution if we are going to introduce more protocols, where the # of existential wrappers grows exponentially.

Comment (sergdort) [reply Bento/Renderable/Deletable.swift]: Why do we need it to `inherit` from `Renderable`?

Comment (andersio) [reply Bento/Renderable/Deletable.swift]: It isn't necessary for `Deletable`. But `Focusable` enforces a `View: FocusableView` constraint on the component.

Comment (andersio) [reply Bento/Renderable/Deletable.swift]: Alternatively, we can drop both `Focusable where Self.View: FocusableView` and `Deletable: Renderable` constraints. Then these protocols would have no associated type and therefore can be used as existentials.

Comment (sergdort) [reply Bento/Renderable/Deletable.swift]: Sounds good 👍 
We can also get rid of `Deletable` inherit from `Renderable`

**Comment by sergdort** [comment]
Also I think we need to add documentation at least to read me that we have this functionality
